### PR TITLE
Add actions to preview editor tab's hover

### PR DESCRIPTION
**Pull request by gjsjohnmurray**
This PR aims to reduce the number of new VS Code users who get confused by preview editors.

It adds `Keep Editor` and `Disable Preview Editors` actions to the hover of a preview editor tab.

image

### CODE REVIEW

**Comment by gjsjohnmurray** [comment]
Pinging @bpasero for initial comments about the concept, and subsequent assignment of a reviewer.

**Comment by benibenj** [comment]
I wonder if this would just lead to new users disabling preview mode instead of educating them on how it works. 
But I'm not sure what the best approach would be to do so. Maybe we could prepend `Preview: ` in the hover. This would allow the user to understand at least how this "mode" is called and would have a way to read about it. We could also link to [documentation](https://code.visualstudio.com/docs/getstarted/userinterface#_preview-mode) about this "mode" (not sure if we do this somewhere else).

**Comment by gjsjohnmurray** [comment]
A quieter alternative that links to the documentation:

image

Text of the hyperlink is italicized to echo how the tab label is displayed.

**Comment by gjsjohnmurray** [comment]
Render the doc link as an icon to make it even quieter:

image

**Comment by gjsjohnmurray** [comment]
ESM changes merged. PR set as draft because most recent experiments (see screenshots above) have so far only been done for multi-tab mode, pending feedback/agreement.

**Comment by benibenj** [comment]
I'm not sure which version looks the best but I think the last (current) one would be a great one to start with. Alternatively to linking to the documentation, we could use the `#(gear)` icon and open the settings editor with the `workbench.editor.enablePreview` setting. The description of the setting includes information on how preview works. 

Screenshot 2024-09-03 155729

**Comment by gjsjohnmurray** [comment]
Ready for review.

image

Hover tip is now element-relative rather than mouse-relative, so it is easier to move the pointer onto. It also now has a 'beak'.

File: src/vs/workbench/browser/parts/editor/editorTabsControl.ts [hunk 1]
```diff
@@ -456,10 +450,6 @@ export abstract class EditorTabsControl extends Themable implements IEditorTabsC
 		return editor.getTitle(Verbosity.LONG);
```
Comment (benibenj) [inline src/vs/workbench/browser/parts/editor/editorTabsControl.ts:450]:
```
protected getHoverTitle(editor: EditorInput): string | IManagedHoverTooltipMarkdownString {
  const title = editor.getTitle(Verbosity.LONG);
  if (!this.tabsModel.isPinned(editor)) {
	  return {
		  markdown: new MarkdownString('', { supportThemeIcons: true, isTrusted: true }).
			  appendText(title).
			  appendMarkdown(' (_preview_ [$(gear)](command:workbench.action.openSettings?%5B%22workbench.editor.enablePreview%22%5D "Configure Preview Mode"))'),
		  markdownNotSupportedFallback: title
	  };
  }
  return title;
}
```

File: src/vs/workbench/browser/parts/editor/singleEditorTabsControl.ts [hunk 2]
```diff
@@ -19,39 +19,98 @@ import { toDisposable } from '../../../../base/common/lifecycle.js';
...
+		private readonly control: SingleEditorTabsControl,
+		@IConfigurationService configurationService: IConfigurationService,
+		@IHoverService hoverService: IHoverService,
+
+	) {
+		super(
+			'element',
+			true,
+			(options) => {
+				return { ...this.getOverrideOptions(options), appearance: { showPointer: true } };
```
Comment (benibenj) [inline src/vs/workbench/browser/parts/editor/singleEditorTabsControl.ts:56]: Regarding `element` & `showPointer`, we tried this about half a year ago but it did not really feel very nice so we reverted that behaviour. I understand that this makes it easier to click the button, but I think users will still press even though it is further away.

**Review by benibenj — CHANGES_REQUESTED** [review]
I think there is a much easier approach to this. We should be able to just add our logic in the `getHoverTitle` method in `editorTabsControl.ts` and everything should work without having to manage multiple hover delegates.

File: src/vs/workbench/browser/parts/editor/editorTabsControl.ts (hunk 1 above)
Comment (gjsjohnmurray) [reply src/vs/workbench/browser/parts/editor/editorTabsControl.ts:450]: Good point! I had originally needed hover delegates so I could add statusbar commands.

File: src/vs/workbench/browser/parts/editor/singleEditorTabsControl.ts (hunk 2 above)
Comment (gjsjohnmurray) [reply src/vs/workbench/browser/parts/editor/singleEditorTabsControl.ts:56]: OK

**Comment by benibenj** [comment]
Thanks for this contribution! Hope this will help new users understand the preview mode. If we notice that this is not enough we can consider other options.

**Review by benibenj — APPROVED** [review]

**Review by bhavyaus — APPROVED** [review]

**Comment by gjsjohnmurray** [comment]
@benibenj please move this to the September 2024 milestone.

**Pull request merged**